From cedbf8a756bdce86a1bbecf80ff823c68ece0da5 Mon Sep 17 00:00:00 2001
From: Andrew Liu <shengping.liu@windriver.com>
Date: Thu, 18 Jun 2009 15:33:46 +0800
Subject: [PATCH 05/25] MPC5121 CAN: Add support for Rev2 silicon

Original patch taken from rev 4 board support ISO image:
  mpc5121ads-20081208_ltib-beta.iso
  http://www.freescale.com/webapp/sps/site/overview.jsp?
  nodeId=0127260061033202A5621E

Rev2 silicon has two additional CAN ports
and more CAN clock source options.

Signed-off-by: Chen Hongjun <hong-jun.chen@freescale.com>
Signed-off-by: John Rigby <jrigby@freescale.com>
Integrated-by: Andrew Liu <shengping.liu@windriver.com>
---
 arch/powerpc/include/asm/mpc512x.h  |    1 +
 drivers/net/can/mscan/mpc52xx_can.c |   39 +++++++++++++++++++++++++++-------
 drivers/net/can/mscan/mscan.h       |    2 +
 3 files changed, 34 insertions(+), 8 deletions(-)

diff --git a/arch/powerpc/include/asm/mpc512x.h b/arch/powerpc/include/asm/mpc512x.h
index 0efc770..9602e4a 100644
--- a/arch/powerpc/include/asm/mpc512x.h
+++ b/arch/powerpc/include/asm/mpc512x.h
@@ -115,6 +115,7 @@ extern void mpc5121_pscgpio_make_psc(int psc, int pin);
 
 extern struct clk *clk_get(struct device *dev, const char *id);
 extern int clk_enable(struct clk *clk);
+extern unsigned long clk_get_rate(struct clk *clk);
 
 #ifdef CONFIG_PM
 extern int __init mpc512x_pm_init(void);
diff --git a/drivers/net/can/mscan/mpc52xx_can.c b/drivers/net/can/mscan/mpc52xx_can.c
index c4c98b7..0d9c6d9 100644
--- a/drivers/net/can/mscan/mpc52xx_can.c
+++ b/drivers/net/can/mscan/mpc52xx_can.c
@@ -48,7 +48,7 @@
 
 RCSID("$Id$");
 
-#define PDEV_MAX 2
+#define PDEV_MAX 4
 
 struct platform_device *pdev[PDEV_MAX];
 
@@ -91,14 +91,24 @@ static int __devinit mpc52xx_can_probe(struct platform_device *pdev)
 	}
 
 	if (pdata->cpu_type == MPC512x_MSCAN) {
-		struct clk *mscan_clk;
-		mscan_clk = clk_get(&pdev->dev, "mscan_clk");
+		struct clk *mscan_clk, *port_clk;
+		char clk_name[15];
+
+		mscan_clk = clk_get(NULL, "mscan_clk");
 		if (!mscan_clk) {
-			dev_err(&pdev->dev, "can't get mscan_clk!");
+			dev_err(&pdev->dev, "can't get mscan_clock!");
 			ret = -EINVAL;
 			goto fail_map;
 		}
 
+		sprintf(clk_name, "mscan%d_clk", pdata->port);
+		port_clk = clk_get(NULL, clk_name);
+
+		/* update clock rate for mpc5121e rev2 chip */
+		if (port_clk)
+			pdata->clock_frq = clk_get_rate(port_clk);
+
+		/* enable clock for mscan module */
 		clk_enable(mscan_clk);
 	}
 
@@ -108,8 +118,8 @@ static int __devinit mpc52xx_can_probe(struct platform_device *pdev)
 
 	ret = register_mscandev(dev, pdata->clock_src);
 	if (ret >= 0) {
-		dev_info(&pdev->dev, "probe for a port 0x%lX done\n",
-			 dev->base_addr);
+		dev_info(&pdev->dev, "probe port 0x%lX done, clk rate:%d\n",
+			 dev->base_addr, pdata->clock_frq);
 		return ret;
 	}
 
@@ -216,10 +226,14 @@ static int __init mpc52xx_of_to_pdev(void)
 	struct device_node *np = NULL;
 	unsigned int i;
 	int ret, type = -1, index = 0;
-	char *mscan_comp_name[] = {"fsl,mpc5200-mscan", "fsl,mpc5121-mscan"};
 	int  cpu_type[] = {MPC52xx_MSCAN, MPC512x_MSCAN};
+	int *port;
+	char *mscan_comp_name[] = {"fsl,mpc5200-mscan",
+				   "fsl,mpc5121rev2-mscan",
+				   "fsl,mpc5121-mscan"};
 
-	for (i = 0; i < 2; i++) {
+
+	for (i = 0; i < 3; i++) {
 		np = of_find_compatible_node(np, NULL, mscan_comp_name[i]);
 		if (np) {
 			type = cpu_type[i];
@@ -265,6 +279,15 @@ static int __init mpc52xx_of_to_pdev(void)
 		pdata.cpu_type = type;
 		pdata.clock_frq = fsl_find_ipb_freq(np);
 
+		if (pdata.cpu_type == MPC512x_MSCAN) {
+			port = (int *)of_get_property(np, "cell-index", NULL);
+			if (!port) {
+				printk(KERN_ERR "Err: can't find can port!\n");
+				goto err;
+			}
+			pdata.port = *port;
+		}
+
 		ret = platform_device_add_data(pdev[i], &pdata, sizeof(pdata));
 		if (ret)
 			goto err;
diff --git a/drivers/net/can/mscan/mscan.h b/drivers/net/can/mscan/mscan.h
index f14e6d6..4c8c4be 100644
--- a/drivers/net/can/mscan/mscan.h
+++ b/drivers/net/can/mscan/mscan.h
@@ -249,6 +249,8 @@ struct mscan_platform_data {
 	u8 clock_src;		/* MSCAN_CLKSRC_BUS or MSCAN_CLKSRC_XTAL */
 	u32 clock_frq;		/* can ref. clock, in Hz */
 	u8 cpu_type;		/* MPC52xx_MSCAN or MPC512x_MSCAN */
+	u8 port;		/* Port number */
+
 };
 
 struct net_device *alloc_mscandev(void);
-- 
1.6.3.1

